toolz/itertoolz.py

```diff
@@ -97,6 +98,32 @@ def groupby(key, seq):
     return rv
 
 
+def indices(*sizes):
+    """ Iterates over a length/shape.
+
+    >>> list(indices(3, 2))
+    [(0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1)]
+
+    This can help nicely index an array.
+
+    >>> l = [[1, 2],
+    ...      [3, 4],
+    ...      [5, 6]]
+
+    >>> for i, j in indices(3, 2):
+    ...     print("l[%i][%i] = %i" % (i, j, l[i][j]))
+    l[0][0] = 1
+    l[0][1] = 2
+    l[1][0] = 3
+    l[1][1] = 4
+    l[2][0] = 5
+    l[2][1] = 6
```

Comment:
I'll try. 😄

So in some cases I have binary data that I need to split up into smaller blocks on in separate processes and potentially combine results from at different stages. This data normally is on disk and may be a single file or split across multiple files. In these cases, I need an index for each block that I will work with. While I suppose one could compute a single index for each block, it makes the code much harder to reason about and it is already somewhat complex code (e.g. adds halos to data blocks, slices out halos afterwards, etc.). Being able to have indices like this makes it easier to reason about these cases and handle arbitrary dimensions. Not to mention stitching the pieces together becomes much more straightforward.

Hopefully that makes sense.